File: delivery/harmony-dingqiao/samples/dingqiao-demo/entry/src/main/ets/util/ColdStartPttGate.ts
export interface ColdStartEngineDecision {
  accepted: boolean;
  finishAfterFlush: boolean;
}

/** Coordinates PTT release, recorder-tail flush, and asynchronous model load. */
export class ColdStartPttGate {
  private generation: number = 0;
  private activeGeneration: number = -1;
  private released: boolean = false;
  private captureDrained: boolean = false;
  private ready: boolean = false;

  begin(): number {
    this.generation += 1;
    this.activeGeneration = this.generation;
    this.released = false;
    this.captureDrained = false;
    this.ready = false;
    return this.activeGeneration;
  }

  release(generation: number): boolean {
    if (generation !== this.activeGeneration) return false;
    this.released = true;
    return true;
  }

  captureStopped(generation: number): boolean {
    if (generation !== this.activeGeneration) return false;
    this.captureDrained = true;
    return this.claimFinish();
  }

  engineReady(generation: number): ColdStartEngineDecision {
    if (generation !== this.activeGeneration) {
      return { accepted: false, finishAfterFlush: false };
    }
    this.ready = true;
    return { accepted: true, finishAfterFlush: this.claimFinish() };
  }

  cancel(): void {
    this.generation += 1;
    this.activeGeneration = -1;
    this.released = false;
    this.captureDrained = false;
    this.ready = false;
  }

  private claimFinish(): boolean {
    if (!this.released || !this.captureDrained || !this.ready) return false;
    this.activeGeneration = -1;
    return true;
  }
}

File: asr/android/samples/dingqiao-demo/src/main/java/com/amphion/dingqiao/demo/ColdStartPttGate.kt
package com.amphion.dingqiao.demo

data class ColdStartEngineDecision(
    val accepted: Boolean,
    val finishAfterFlush: Boolean,
)

/** Coordinates PTT release, recorder-tail flush, and asynchronous model load. */
class ColdStartPttGate {
    private var generation = 0
    private var activeGeneration = -1
    private var released = false
    private var captureDrained = false
    private var ready = false

    @Synchronized
    fun begin(): Int {
        generation += 1
        activeGeneration = generation
        released = false
        captureDrained = false
        ready = false
        return activeGeneration
    }

    @Synchronized
    fun release(requestGeneration: Int): Boolean {
        if (requestGeneration != activeGeneration) return false
        released = true
        return true
    }

    @Synchronized
    fun captureStopped(requestGeneration: Int): Boolean {
        if (requestGeneration != activeGeneration) return false
        captureDrained = true
        return claimFinish()
    }

    @Synchronized
    fun engineReady(requestGeneration: Int): ColdStartEngineDecision {
        if (requestGeneration != activeGeneration) {
            return ColdStartEngineDecision(accepted = false, finishAfterFlush = false)
        }
        ready = true
        return ColdStartEngineDecision(accepted = true, finishAfterFlush = claimFinish())
    }

    @Synchronized
    fun cancel() {
        generation += 1
        activeGeneration = -1
        released = false
        captureDrained = false
        ready = false
    }

    private fun claimFinish(): Boolean {
        if (!released || !captureDrained || !ready) return false
        activeGeneration = -1
        return true
    }
}
